### ISSUE TITLE
C sharp wrapping should use a text template 

### ISSUE THREAD

**Issue by Laeeth** [kaleidicassociates/autowrap#45]
It's hard to follow.

Either use poor man's substitute a la Atila /string interpolation or some kind of template like vibe diet.  Former better for compile times.

**Comment by LightBender** [comment]
I have very strong opinions about text templates as a result of my time at MSFT. MSFT has a very good TT system called T4, and early in the design phase of our project it was decided that we would use T4 to generate our queries. At the time I left 3 years later it was universally (even by the architect who originally pushed for it) considered to be the worst design decision we ever made. They were effectively untestable and the tests we did have required constant manual upkeep. They were incredibly difficult to maintain and we had to build an entire metasystem to be able to actually use them for anything. The problems were considered so bad that an entire engineer's output over the course of the last two years was soaked up building a replacement for it, however, given the pervasiveness of the templates and subsequent scope of work, that project had not yet been completed at the time I left. I was the most vocal engineer against T4 as I was only person on the team with significant experience in code-generation at the time, but MSFT being the hierarchical organization that it is, I was overruled by the architect. The architect later admitted that I was right but by that time it was to late to go back.

To paraphrase an ancient programmer joke: Some people, when confronted with a code generation problem, think "I know, I'll use text templates." Now they have two problems.

TT systems shine when you are a replacing a small number of tokens within large blocks of text. ASP.NET's Razor or Vibe's Diet (which are both intended to achieve the same result) are great at what they do because in a web page you usually are only replacing a small number of tokens within a large amount of boilerplate. And indeed I actually use that model in the Ranges.cs and Boilerplate.cs files.

Consider the C# template for a D module function:
[DllImport("%1$s", EntryPoint = "%2$s", CallingConvention = CallingConvention.Cdecl)]
private static extern %3$s %4$s(%5$s);
public static %6$s %7$s(%8$s) {
    var dlang_ret = %9$s;
    return %10$s;
}

I have a rule-of-thumb for templates versus inline generation. If the amount of boilerplate text is less than half of the total expanded text, then use inline generation, otherwise use a template. Now in the actual implementation the first line IS a template, and even then, the full example above always expands to more than 50% of the templates initial size.

I absolutely must recommend in the strongest possible terms that we NOT attempt to use text templates.

**Comment by LightBender** [comment]
I'm perfectly OK with doing this work, I just wanted offer up my prior experience before committing to this course of action.

**Comment by atilaneves** [comment]
I recommend `scriptlike.interp` for this kind of thing.

**Comment by LightBender** [comment]
Hmmm, I'd never heard of that one. I'll check it out.
